src/Avalonia.Base/AvaloniaProperty`1.cs

```diff
@@ -38,8 +42,31 @@ public abstract class AvaloniaProperty<TValue> : AvaloniaProperty
             PropertyMetadata metadata)
             : base(source, ownerType, metadata)
         {
+            _changed = source is AvaloniaProperty<TValue> p ? p._changed : new Subject<AvaloniaPropertyChangedEventArgs<TValue>>();
```

Comment:
This doesn't make much sense for untyped `AvaloniaProperty`, but is there actually a good alternative?

Reply:
I think we can throw if `source` is not of the correct type - there shouldn't be a case where a wrongly-typed `AvaloniaProperty` gets here, and if it does then other things will break I think?

Reply:
I replaced the copy constructors for `AvaloniaProperty<TValue>` and `DirectPropertyBase<TValue>` to require same property type, and marked the old ones as obsolete with error.